src/request.ts

```diff
@@ -0,0 +1,362 @@
+/**
+ * Type safe wrapper around the fetch API
+ */
+
+// TODO: add body type parameter to POST/PUT
+// TODO: support for optional query parameters
+// TODO: when query/headers type is "never", it should not allow any query/header to be produced
+// TODO: add timeout to fetch (how to cancel request?)
+// TODO: add etag support in responses
+
+// The type definition of fetch is needed because it's not provided by TypeScript
+export declare function fetch(
+  input: RequestInfo,
+  init?: RequestInit
+): Promise<Response>;
+
+import * as t from "io-ts";
+
+/**
+ * Describes the possible methods of a request
+ */
+export type RequestMethod = "get" | "post" | "put";
+
+/**
+ * Describes the possible header keys of a request
+ */
+// tslint:disable-next-line:no-duplicate-string
+export type RequestHeaderKey = "Authorization" | "Content-Type";
+
+/**
+ * Describes a set of headers whose keys are of type RequestHeaderKey
+ */
+export type RequestHeaders<HS extends RequestHeaderKey> = {
+  [key in HS]: string
+};
+
+/**
+ * Describes the query params for this request
+ */
+export type RequestQuery<K extends string> = Record<K, string>;
+
+/**
+ * Generates a set of headers with certain keys (KH) from a parameters object
+ * of type P.
+ */
+export type RequestHeaderProducer<P, KH extends RequestHeaderKey> = (
+  params: P
+) => RequestHeaders<KH>;
+
+/**
+ * Composes two RequestHeaderProducer(s)
+ */
+export function composeHeaderProducers<
+  PI,
+  KI extends RequestHeaderKey,
+  PH,
+  KH extends RequestHeaderKey
+>(
+  p0: RequestHeaderProducer<PI, KI>,
+  p1: RequestHeaderProducer<PH, KH>
+): RequestHeaderProducer<PI & PH, KI | KH> {
+  return params => {
+    const headers0 = p0(params);
+    const headers1 = p1(params);
+    return {
+      // tslint:disable-next-line:no-any
+      ...(headers0 as any),
+      // tslint:disable-next-line:no-any
+      ...(headers1 as any)
+    };
+  };
+}
+
+/**
+ * Describes a possible response type: when the status code is S, the response
+ * type is T.
+ */
+export interface IResponseType<S extends number, T> {
+  readonly status: S;
+  readonly value: T;
+}
+
+/**
+ * A function that generates a typed representation of a response.
+ * It should return undefined in case the response cannot be decoded (e.g.
+ * in case of a parsing error).
```

Comment:
why not an either with the error message ?

Reply:
I initially did that but then I decided to make the API simpler, not sure it's actually useful in the context of our APP.